JS refactor - remove dependency on Google Closure Library

bensampaio opened the pull request:
I rewrote the all JS with ES2015 syntax and used Jasmine for testing the Router class. This way there is no need to depend on external libraries. I also added the possibility of including the Router class and Routing instance as a module instead of global (but for this the package should ideally be published on npm).

This new version should be compatible with older versions.

For future development:
- First make sure you have Node 6+
- Then on a terminal go to the `Resources` folder;
- Run `npm install`;
- To transpile the `Resources/js/router.js` run `npm run build` (generated files will be put in `Resources/public/js`);
- To test run `npm test`;

stof commented on Resources/js/router.js:
```diff
@@ -225,45 +244,91 @@ fos.Router.prototype.generate = function(name, opt_params, absolute) {
+ *     hosttokens: (Array.<string>)
+ * }}
+ */
+Router.Route;
```
we don't need to expose it (it is not exposed in the old code)

bensampaio replied:
I believe this is fixed when using the UMD template, or am I missing something?

stof commented on Resources/js/router.js:
```diff
@@ -225,45 +244,91 @@ fos.Router.prototype.generate = function(name, opt_params, absolute) {
+ *     base_url: (string)
+ * }}
+ */
+Router.Context;
```
same here

bensampaio replied:
I believe this is fixed when using the UMD template, or am I missing something?

stof commented on Resources/js/router.js:
```diff
@@ -1,222 +1,241 @@
+     * @returns {Router}
+     */
+    static getInstance() {
+        return instance || new Router();
```
this is broken. It will not always return the same instance

stof commented on Resources/js/router.js:
```diff
@@ -1,222 +1,241 @@
+     * @param {Object.<string, Router.Route>=} routes
+     */
+    constructor(context, routes) {
+        instance = this;
```
this is wrong. The old code was still allowing to create multiple instance of the router, when using the constructor rather than using the singleton getter.. The router itself is not a singleton

stof commented on Resources/package.json:
```diff
@@ -0,0 +1,41 @@
+    "public/js/router.min.js",
+    "ts/router.d.ts"
+  ],
+  "main": "js/router.js",
```
shouldn't the main file be the ES5 one, so that people don't need to preprocess their node modules with babel ?

stof commented:
You compiled file puts the global namespace in strict mode. This is a total no-go

stof commented:
and btw, it also leaks many variables when including it in the page instead of using a CommonJS loader. You need to make it support non-CommonJS environment properly

stof commented:
btw, it would be great to send a PR rewriting the JS **without** rewriting the testsuite. This way, it is much easier to be sure it does not break things

stof commented on Resources/public/js/router.js:
```diff
@@ -1,12 +1,400 @@
+        value: function setData(data) {
+            var router = Router.getInstance();
+
+            router.setBaseUrl( /** @type {string} */data['base_url']);
```
all these ``@type`` are useless once you don't use the Google closure compiler (we were running it in strict mode)

stof commented on Resources/public/js/router.js:
```diff
@@ -1,12 +1,400 @@
+ * @type {null|Router}
+ */
+
+var _typeof = typeof Symbol === "function" && typeof Symbol.iterator === "symbol" ? function (obj) { return typeof obj; } : function (obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol ? "symbol" : typeof obj; };
```
why does the build include a Symbol polyfill while we don't use symbols ? this looks wrong to me. We should not bundle a useless polyfill in our file

stof commented:
and Travis is broken as you have not updated the way to run JS tests...

javiereguiluz commented:
@bensampaio this looks really promising! Thank you.

bensampaio commented:
Thanks for the feedback. I made the following changes:

- Removed 'use strict' from the global scope
- Exposed only the necessary variables (`Router` and `Routing`), is that correct?
- Removed unnecessary `@type`s
- Fixed the `getInstance()` method
- Fixed the main file
- Tried to fix Travis but it seems I failed... Can you give me some light here? Never worked with Travis and I can't link it to my repo.

The `_typeof` function you see in the top of the file is added by Babel. Every babel project always has that function in the top. It isn't a `Symbol` polyfill it is a function to detect `Symbol` types correctly. Useless for this case I admit, but I couldn't find a way to remove it. In the end you still get a much smaller minified file than before and you don't depend on an external lib.

Regarding the test suite change, the previous version was written in jsunit which I didn't hear about in years. So I decided to go looking and found this on the repo description: "The original unit-testing framework for JavaScript. These days we use Jasmine (http://github.com/pivotal/jasmine) by default for JS testing; JsUnit is not actively developed or supported" (https://github.com/pivotal/jsunit). Since it is no longer supported and since I believe in the powers of Jasmine I decided to make the change :) In any case, the tests are exactly the same as before. They are just grouped differently and the assert functions were renamed to their equivalent in jasmine.

stof commented:
@bensampaio but the rewrite of tests at the same time makes it hard to know whether the testsuite is equivalent or no. I suggest keeping the testsuite rewrite for a separate PR (which would not change the codebase but only tests)

stof commented on Resources/public/js/router.js:
```diff
@@ -1,12 +1,407 @@
+
+var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticP ...
+
+function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }
```
you generated file still defines lots of global variables. This is a no-go.

bensampaio replied:
I'm trying to find a solution for this. It seems babel always creates those variables in the global scope which is a bit weird and unexpected.

stof replied:
I think the solution is not to use babel as the only build tool. We should enclose the file in a UMD boilerplate *after* transpiling it with Babel.
If you don't work on it, I may try after my vacations

bensampaio replied:
I found a solution I believe :)

I used the returnExports UMD template to export the Router and Routing variables only. To make sure babel variables don't polute the global scope I used gulp to transpile the ES2015 code, then insert the result into the UMD template and then uglify it.

stof commented:
@bensampaio FYI, I will be in vacation for the next 2 weeks (starting tomorrow), so further review on this PR will have to wait until I come back.

stof commented on .travis.yml:
```diff
@@ -33,8 +33,8 @@ before_install:
 install:
   - composer install
-  - npm install google-closure-library
+  - cd Resources && npm install
```
running phpunit is now broken, because of the ``cd Resources`` you added here, as you don't come back to the root after that

bensampaio commented:
@stof I understand what you mean about the tests. However, I only rewrote the tests because I believed that would be a great and welcome improvement. And I can assure you the tests are exactly the same as before just running under a different framework.

In any case, I can put the old tests back but I won't have the time to work on a PR just for tests improvement. Mainly because the tests are not my priority and undoing and redoing in a separate branch will still take sometime besides all the time required to actually get the PR accepted.

If you need more time to confirm the tests are actually equivalent then fine. If that's not ok then I'll remove them.

bensampaio commented:
What is the state on this?

stof commented:
@bensampaio I came back from vacation a few days ago, and I haven't done any programming work during the weekend. So there has been no progress yet since the end of July.
I want to look at it in more details before merging it (for instance checking that the tests are equivalent).

tcitworld commented:
Any news ?

yellow1912 commented:
Another friendly ping ^_, I think it's really promising.

raulferras commented on Resources/js/router.template.js:
```diff
@@ -0,0 +1,23 @@
+(function (root, factory) {
+    if (typeof define === 'function' && define.amd) {
+        // AMD. Register as an anonymous module.
+        define(['fos'], factory);
```
If I'm not mistaken, here you are specifying that this module requires a module called `fos`.
If your intention was to give a name to this module, you should use a plain string. However, that is something not recomended (naming your module).
I would just do:

`define([], factory);`

weaverryan commented:
bump. @stof is this just waiting on you for a review? Or could I also help out? 

@bensampaio is the usage the same after this? i.e. I could still just add the 2 script tags (for the Router & for the routes)? And if I `require` this as a module, what does that look like? How do we pass the routes in?

I'd like this library to be friendly with Webpack/Encore

willdurand commented:
It is tricky because tests have been rewritten and I am personally afraid to introduce bugs. Very first version of this bundle used standard JS and then it has been rewritten. Now we are basically screwed 😞 

I would like to keep the original test suite to ensure the rewritten lib is compliant, and maybe rewrite the test suite later. This test suite is quite comprehensive, so I really trust it.

bensampaio commented:
@weaverryan after this you should still be able to use this as before but also as a module if that fits your use case better.

If you need this feel free to take over my work.

stof commented:
Well, the fact that both tests and code are rewritten at the same time makes the review really hard (and I admit I forgot about this PR).

isychev commented:
If someone will be interested, I make npm package [fos-routing](https://github.com/isychev/fos-routing) It's a standard `Resources/js/router.js`

tobias-93 commented:
Closing in favor of #303
